test/terra/backends/qasm_simulator/qasm_snapshot.py

```diff
@@ -0,0 +1,338 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+QasmSimulator Integration Tests for Snapshot instructions
+"""
+
+import numpy as np
+
+from qiskit.compiler import assemble
+from qiskit.quantum_info.operators import Pauli
+from qiskit.providers.aer import QasmSimulator
+from qiskit.providers.aer import AerError
+
+from test.terra.reference.ref_snapshot_state import (
+    snapshot_state_circuits_deterministic, snapshot_state_counts_deterministic,
+    snapshot_state_pre_measure_statevector_deterministic,
+    snapshot_state_post_measure_statevector_deterministic,
+    snapshot_state_circuits_nondeterministic,
+    snapshot_state_counts_nondeterministic,
+    snapshot_state_pre_measure_statevector_nondeterministic,
+    snapshot_state_post_measure_statevector_nondeterministic)
+
+
+class QasmSnapshotStatevectorTests:
+    """QasmSimulator snapshot statevector tests."""
+
+    SIMULATOR = QasmSimulator()
+    SUPPORTED_QASM_METHODS = [
+        'automatic', 'statevector', 'matrix_product_state'
+    ]
+    BACKEND_OPTS = {}
+
+    def statevector_snapshots(self, data, label):
+        """Format snapshots as list of Numpy arrays"""
+        # Check snapshot entry exists in data
+        self.assertIn("snapshots", data)
+        self.assertIn("statevector", data["snapshots"])
+        self.assertIn(label, data["snapshots"]["statevector"])
+        # Format output as list of numpy array
+        snaps = data["snapshots"]["statevector"][label]
+        statevecs = []
+        for snap in snaps:
+            svec = np.array(snap)
+            statevecs.append(svec[:, 0] + 1j * svec[:, 1])
+        return statevecs
+
+    def test_snapshot_statevector_pre_measure_det(self):
+        """Test snapshot statevector before deterministic final measurement"""
+        shots = 10
+        label = "snap"
+        counts_targets = snapshot_state_counts_deterministic(shots)
+        statevec_targets = snapshot_state_pre_measure_statevector_deterministic(
+        )
+        circuits = snapshot_state_circuits_deterministic(label,
+                                                         'statevector',
+                                                         post_measure=False)
+
+        qobj = assemble(circuits, self.SIMULATOR, shots=shots)
+        job = self.SIMULATOR.run(qobj, backend_options=self.BACKEND_OPTS)
+        method = self.BACKEND_OPTS.get('method', 'automatic')
+        if method not in QasmSnapshotStatevectorTests.SUPPORTED_QASM_METHODS:
+            self.assertRaises(AerError, job.result)
```

Comment:
Can this really happen? As we are setting the `BACKEND_OPTS` to `{}`, the `method` variable will always contain the default value `automatic`, isn't it?

Reply:
This is set by the test classes themselves, and checks that the unsupported instructions are caught and raise exception correctly (eg. stabilizer snapshot on density matrix method)